internal/db/schema/migrations/oss/postgres/70/02_sessions.up.sql

```diff
@@ -0,0 +1,189 @@
+begin;
+
+  -- Replaces trigger from 56/02_add_data_key_foreign_key_references.up.sql
+  drop trigger cancel_session_with_null_fk on session;
+  create or replace function cancel_session_with_null_fk() returns trigger
+  as $$
+  begin
+    -- Note that we need each of these to run in case
+    -- more than one of them is null.
+    if new.auth_token_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    if new.project_id is null then
+      -- Setting the key_id to null will allow the scope
+      -- to cascade delete its keys.
+      new.key_id = null;
+      perform cancel_session(new.public_id);
+    end if;
+    if new.target_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    if new.user_id is null then
+      perform cancel_session(new.public_id);
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+  
+  create trigger cancel_session_with_null_fk before update of auth_token_id, project_id, target_id, user_id on session
+    for each row execute procedure cancel_session_with_null_fk();
+
+  create table session_target_address (
+    session_id wt_public_id primary key,
+    target_id wt_public_id,
+    -- the following foreign key constraint is set to null on a delete
+    -- because we want to be able to invoke a trigger to call cancel_session
+    -- when the references column is deleted
+    -- rather than canceling a session for any reason that may cause a row to be deleted.
```

Comment:
Is it valid to insert a row with a null target_id? If that is an invalid state, can we add a `before insert` trigger to prevent it?

Reply:
It is not valid to insert a row with a null target_id. In the session domain, we return an error to the user if target_id is empty when trying to create a SessionTargetAddress object in memory. I think it's a good idea to add the trigger your are suggesting. Would you like me to do the same thing for session_host_set_host?

Reply:
Yes, that sounds good.